test/special-char-salt.c

```diff
@@ -0,0 +1,848 @@
+/* Copyright (C) 2020 Björn Esser <besser82@fedoraproject.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/* Simple test to ensure compatability of our implementations of the
+   md5crypt, sha256crypt, and sha512crypt hashing method with the ones
+   found in other commonly used libcrypt libraries.  Their implementations
+   allow for any ASCII character (with one exception to the colon ':'
+   character, that has special meaning in the Unix shadow file) to be
+   present in a salt string.  */
+
+#include "crypt-port.h"
+
+#if INCLUDE_md5crypt || INCLUDE_sha256crypt || INCLUDE_sha512crypt
+
+#include <stdio.h>
+
+struct testcase
+{
+  const char *setting;
+  const char *expected;
+};
+
+static const struct testcase testcases[] =
+{
+#if INCLUDE_md5crypt
+  {"\x24\x31\x24",                                                                 "\x24\x31\x24\x24\x4e\x78\x72\x72\x2e\x58\x54\x69\x53\x36\x41\x59\x2e\x37\x53\x32\x79\x65\x33\x4f\x41\x31"},
+  {"\x24\x31\x24\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01", "\x24\x31\x24\x01\x01\x01\x01\x01\x01\x01\x01\x24\x72\x76\x74\x56\x6e\x2f\x4d\x4c\x77\x52\x2e\x37\x33\x55\x4e\x6b\x33\x79\x41\x78\x43\x30"},
```

Comment:
I think these table entries would be easier to read if _only_ the salt was expressed with hex escapes:

```
  {"$1$",                                                                 "$1$$Nxrr.XTiS6AY.7S2ye3OA1"},
  {"$1$\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01", "$1$\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01\x01$rvtVn/MLwR.73UNk3yAxC0e"},
  // etc
```

Reply:
Alrighty, let me rework this.

Reply:
Fixed in rebased commit.